### ISSUE TITLE
[enum.udecl] Unclear handling for `using enum` with local enums in function templates

### ISSUE THREAD

**Issue by brevzin** [cplusplus/CWG#113]
Full name of submitter (unless configured in github; will be published with the issue): Sean Baxter

Reference (section label): [enum.udecl] and [temp.dep.type]

Issue description:

Consider:

```cpp
template <typename T>
void func() {
    enum class foo_t {
        circle, square
    };

    using enum foo_t;
}
```

Is this valid?

`foo_t` doesn't meet any of the conditions of a dependent type as laid out in [\[temp.dep.type\]/7](https://eel.is/c++draft/temp.dep.type#7), but probably should count as dependent. Once it does, we have this rule in [\[enum.udecl\]/1](https://eel.is/c++draft/enum.udecl#1):

> The [elaborated-enum-specifier](https://eel.is/c++draft/dcl.type.elab#nt:elaborated-enum-specifier) shall not name a dependent type and the type shall have a reachable [enum-specifier](https://eel.is/c++draft/dcl.enum#nt:enum-specifier)[.](https://eel.is/c++draft/enum.udecl#1.sentence-1)

gcc rejects the above example. But the intent in rejecting the dependent case was more like the following, where we can't know what the names of the enumerators are:

```cpp
template <typename T>
void bad() {
    using enum T::E;
}
```

The original example, we have them: `circle` and `square`. That case should probably be allowed - `foo_t`'s underlying type could be dependent, as could the values of `circle` and `square`, but not the actual names themselves.

**Comment by seanbaxter** [comment]
There's a second wording problem in [enum.udecl].
```cpp
enum class foo_t {
  a, b, c
};

using MyEnum = foo_t;

int main() {
  using enum MyEnum;
}
```
https://godbolt.org/z/e7zxq38fj

Does the _using-enum-declaration_ use an _elaborated-type-specifier_ lookup here? Clang thinks it does, so MyEnum is simply not found, as an alias doesn't match an _enum-specifier_. gcc/msvc both find alias.

**Comment by jensmaurer** [comment]
Barry, I think your initial example is valid per the wording and clang accepts it. Please file a bug report for gcc. `foot_t`, as used in `func`, is obviously not dependent (and there seems to be no reason why it ought to be dependent).

The example from @seanbaxter is more interesting, since [enum.udecl] does not use the grammar non-terminal _elaborated_type_specifier_. I've created CWG2621 for that.

**Comment by seanbaxter** [comment]
> foot_t, as used in func, is obviously not dependent 

`foo_t` is dependent, but it's a member of the current instantiation, so its enumerators should be accessible. I think the wording should be fixed.

**Comment by jensmaurer** [comment]
The term "current instantiation" does not apply to function templates, only to class templates, it seems. See [temp.dep.type] p1. Do you have a suggestion how to "fix" the wording?

Given the status quo wording, why do you believe `foot_t` is dependent? A local class in its position is also not dependent, per the wording.

**Comment by JohelEGP** [comment]
> (The idea of "dependent" is that different template specializations could refer to different entities. That's not the case for local enums and local classes inside function templates.)

Implementations suggest otherwise: https://godbolt.org/z/hz6csfabe.

**Comment by cmeerw** [comment]
related issue: [CWG 2074](https://cplusplus.github.io/CWG/issues/2074.html)

**Comment by jensmaurer** [comment]
Amended CWG2074 with the local enum situation; the treatment of local enums and local classes should be consistent.

**Comment by seanbaxter** [comment]
> The term "current instantiation" does not apply to function templates, only to class templates, it seems. 

```cpp
template <typename T>
struct foo_t {
  enum class E {
    circle, square
  };

  using enum E;
};
```
https://godbolt.org/z/nEevxK9se

The same problem occurs with an enum in a class template, which is definitely a member of the current instantiation. Maybe reword like this:

> The _elaborated-enum-specifier_ shall not name a dependent type, unless the type is a member of the _current instantiation_ or surrounding templated function.

Any class/enum type that is defined where template parameters are visible is a dependent type. I wish we could expand "current instantiation" to apply to templated functions as well as classes.

**Comment by jensmaurer** [comment]
> I wish we could expand "current instantiation" to apply to templated functions as well as classes.

I suspect that will be the more consistent approach. Note that we need to cover local classes, too.
